### ISSUE TITLE
Respect termination signals

### ISSUE THREAD

**Issue by TiloW** [frerich/clcache#261]
[`clcache` ignores termination signals](https://github.com/frerich/clcache/blob/4b0ebec4de5dd55f19c1597a555c4f111e001d29/clcache.py#L1456-L1459).
This is bad because a single call to `cl` may take rather long. For example, when using large header-only libraries or invoking the compiler for many source files at once.

**Comment by frerich** [comment]
The code was originally merged as part of #233, which addresses a real issue. Maybe a compromise would be to adjust the code such that the signals are only ignored while the cache is modified?

**Comment by TiloW** [comment]
> a compromise would be to adjust the code such that the signals are only ignored while the cache is modified

Either that or handle corrupt cache entries correctly, i.e., invalidate them upon lookup.
More precisely, these signals should never be ignored but one might decide to wait for the cache modification to finish until exiting the program.

**Comment by siu** [comment]
@TiloW it was important for us to ignore the signals as it was too common that we aborted compilation with ctrl-c and the cache would be corrupted.

A possible improvement is to limit the time that the signals are ignored. In the end what really needs to be protected are the operations that persist data to disk:

- copyOrLink of the compiler artifacts
- writing manifest entries
- writing the statistics
- what else?

**Comment by frerich** [comment]
How about a context manager for this, e.g. (untested)

```python
@contextlib.contextmanager
def blockedSignals(signals=None):
    signals = signals or [signal.SIGINT, signal.SIGTERM]

    handlers = [signal.signal(sig, signal.SIG_IGN) for sig in signals]
    try:
        yield
    finally:
        for sig, handler in handlers:
            signal.signal(sig, handler)
```

With this at hand, one could wrap relevant sections of code as:

```python
with blockedSignals():
    doWriteData()
    doWriteMoreData()
    ...
```

**Comment by siu** [comment]
Looks like a good solution and it doesn't make the code that uses it much harder to understand. :+1:

**Comment by TiloW** [comment]
>  it was important for us to ignore the signals as it was too common that we aborted compilation with ctrl-c and the cache would be corrupted

This is indeed a problem. We're running builds that take several minutes and call the compiler exactly twice. So basically, once started, we can't cancel the build. Additionally, there is the fact that the OS might decide to terminate the compiler anyways (for whatever reasons). It would be nice to have a solution that is able to handle such cases as well, i.e., recognize corrupt cache files.

**Comment by TiloW** [comment]
Would it be possible do add some sort of zero bytes file that is created only after all modifications to the respective cache entry are done and deleted prior to modifying anything? So that we can easily check, just by the existence oft this file, whether we have in fact a valid cache entry.

**Comment by frerich** [comment]
> Would it be possible do add some sort of zero bytes file that is created only after all modifications to the respective cache entry are done and deleted prior to modifying anything?

That's an interesting idea. Another (crazy?) idea which came up some time ago was to rename the cache directory while the entry is being edited, e.g. by appending ` -dirty` prefix or so.

For what it's worth, I'm personally more concerned about the question: in case we notice that a cache entry is invalid, what to do about it? After all, an invalid entry in the compiler artifact repository might be referenced by a lot of manifests, too. So there's some sort of "rippling effect".

**Comment by frerich** [comment]
In any case, I think this is not an either-or thing: blocking signals only while changes are written to disk probably makes sense, and is in scope for this ticket. Making clcache handle broken caches more gracefully would be a topic for a separate ticket.

**Comment by TiloW** [comment]
> blocking signals only while changes are written to disk probably makes sense, and is in scope for this ticket

Certainly an improvement but doesn't resolve this issue as the signals are still ignored (sometimes).

> in case we notice that a cache entry is invalid, what to do about it? 

I'm probably missing something here: why can't we just re-compile it?

**Comment by frerich** [comment]
The problem you describe is not that the signals are ignored, is it? The issue is that they are ignored for most of the lifetime of the clcache process, and that issue manifests as a problem for your use case because clcache is invoked with a lot of source files and thus runs for a long time - sounds accurate?

If so, then I imagine that merely shortening the phases during which the signals are blocked would resolve the problem, no?

**Comment by frerich** [comment]
> I'm probably missing something here: why can't we just re-compile it?

Maybe we can - the code would just need to detect all cases where something could have gone wrong (say: corrupt JSON files or missing `object` files or so) and then signal some 'malformed cache entry' condition. And when handling that condition, the code should not break because there are (at least partially) existing files already.

**Comment by TiloW** [comment]
> I imagine that merely shortening the phases during which the signals are blocked would resolve the problem

Actually, I don't know whether that would be sufficient for my use case. We're using clcache for CI with [gitlab multi runner](https://gitlab.com/gitlab-org/gitlab-ci-multi-runner) and sometimes there are certain builds that need to be canceled. I don't know (and don't want to make any assumption about) how cancellation is implemented. We don't know whether a second `SIGTERM` is fired if the first one did not succeed (so the change proposed above could be problematic) nor whether the OS simply terminates the process. Corrupting the cache during cancellation is obviously bad.

In conclusion, a better title for this issue would have been "Prevent cache corruption". Should I open another one? Even though I believe that when speaking about robust solutions this boils down to the same issue.

**Comment by frerich** [comment]
> In conclusion, a better title for this issue would have been "Prevent cache corruption". Should I open another one?

Indeed, as I mentioned earlier - I don't think these are mutually exclusive options. It would be great to make clcache try hard to avoid writing out broken cache data (e.g. by ignoring `SIGTERM` and `SIGINT` in certain phases) *and* make it handle defective caches gracefully. After all, a cache may have become corrupted by external sources.

So a second ticket would make perfect sense to me, but I probably wouldn't call it "Prevent cache corruption" -- because we cannot prevent it. Instead, it would be more like "Gracefully handle corrupt caches". Sounds good?

**Comment by siu** [comment]
> Actually, I don't know whether that would be sufficient for my use case. We're using clcache for CI with gitlab multi runner and sometimes there are certain builds that need to be canceled. I don't know (and don't want to make any assumption about) how cancellation is implemented. We don't know whether a second SIGTERM is fired if the first one did not succeed (so the change proposed above could be problematic) nor whether the OS simply terminates the process. Corrupting the cache during cancellation is obviously bad.

Would make sense to instead of ignoring the signals delay the handling until it is safe to do so? I.e. install a signal handler that handles the interruption after the protected code is finished?

**Comment by TiloW** [comment]
> Would make sense to instead of ignoring the signals delay the handling until it is safe to do so

Sounds good. That would be more robust than the initially suggested change and prevents cache corruption as long as the compiler is not terminated without sending a signal first (assuming the protected blocks don't take too long to execute).

**Comment by frerich** [comment]
> Would make sense to instead of ignoring the signals delay the handling until it is safe to do so? I.e. install a signal handler that handles the interruption after the protected code is finished?

Not sure I get this - can you elaborate? In which phase(s) do you want to delay signal handling, and until when do you want to delay them?

**Comment by TiloW** [comment]
I believe @siu means something like this:

```python
@contextlib.contextmanager
def blockedSignals(signals=None):
  signals = signals or [signal.SIGINT, signal.SIGTERM]
  
  doTerminate = false
  def onTerminationRequested:
    doTerminate = true

  try:
    handlers = [signal.signal(sig, onTerminationRequested) for sig in signals]
    yield
  finally:
    if doTerminate:
      sys.exit()
    for sig, handler in zip(signals, handlers):
      signal.signal(sig, handler)
```

**Comment by siu** [comment]
Yes, that is what I had in mind. With that solution the signals will be honored but only when it is safe.

**Comment by frerich** [comment]
Ah, I see -- I fear this suffers from the same issue though as the more primitive version I proposed in #262 -- `signal.signal` can only be called in the main thread, so the `blockedSignals` call would need to be done on a fairly 'outer' scope -- effectively keeping the signals ignored while the compiler(s) is/are running (which maybe be for a long time when invoked with multiple source files).

**Comment by TiloW** [comment]
> signal.signal can only be called in the main thread

I don't know much about python but I'm pretty sure this can still be accomplished. Handlers would always be installed but exiting could be delayed depending on the current state. We need some sort of context manager that communicates with the main thread (simply via thread locks?) and registers the thread as currently executing protected code. After receiving a termination request, the program is shutdown only if there is no protected execution going on. Otherwise, we wait until protected execution finishes and exit afterwards. Thread registration needs to be rejected when termination was requested before.

**Comment by TiloW** [comment]
More precisely, there would be a global `protectedThreadCounter = 0` and `terminationWasRequested = False`.

And a context manager, like so
```python
@contextlib.contextmanager
def protectThread:
  try:
    if terminationWasRequested:
      sleep # indefinitely
    protectedThreadCounter += 1
    yield
  finally:
    protectedThreadCounter -= 1
```

And a signal handler in the main thread:
```python
def onTerminationRequested:
  if not terminationWasRequested:
    terminationWasRequested = True
    while protectedThreadCounter > 0:
      sleep 1
    sys.exit
signal.signal(sig, onTerminationRequested) for sig in [signal.SIGINT, signal.SIGTERM]
```

Not sure if `protectedThreadCounter += 1` is an atomic operation in python, we might need locks.

**Comment by frerich** [comment]
Maybe it would indeed be easier to drop the idea of handling signals gracefully and rather

1. Do a better job at handling corrupted caches
2. Avoid that caches get written to (or read from) partially by writing to temporary file and then using something like [`ReplaceFile`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365512(v=vs.85).aspx).

**Comment by TiloW** [comment]
> Do a better job at handling corrupted caches

That would indeed be the most robust solution. Depending on the performance-impact of recompiling corrupted cache entries it could still be beneficial to include the above change.

**Comment by frerich** [comment]
> That would indeed be the most robust solution.

Alas, it would not solve all issues addresses by ignoring the signals; in particular, I seem to recall that one issue was that signals which occur while writing to the cache or reading from it might cause a *partial* file copy. So e.g. a cache entry was added but copying the object file into it was interrupted -- this would not get detected, and subsequent compiled would restore a broken object file.

**Comment by TiloW** [comment]
> signals which occur while writing to the cache or reading from it might cause a partial file copy

Well, this could also be handled depending on how cache entries are validated (e.g., by removing a modification-marker-file only after copying is finished).

**Comment by frerich** [comment]
Closed by merging #298
